### PR TITLE
Get arithmetic operation kind for compound assignments

### PR DESCRIPTION
**Pull request by d367wang**
A split of #331 

1. Get arithmetic operation kind for compound assignments
2. Small tweaks on parameters of `SlotManager.createArithmeticVariableSlot`

Co-authored-by: Jenny Xiang [j.tt.xiang@gmail.com](mailto:j.tt.xiang@gmail.com)

### CODE REVIEW

File: src/checkers/inference/DefaultSlotManager.java [hunk 1]
```diff
@@ -601,7 +598,7 @@ private TypeKind getArithmeticResultKind(AnnotatedTypeMirror lhsAtm, AnnotatedTy
 
     @Override
     public ArithmeticVariableSlot createArithmeticVariableSlot(
-            AnnotationLocation location, AnnotatedTypeMirror lhsAtm, AnnotatedTypeMirror rhsAtm) {
+            AnnotationLocation location, TypeMirror lhs, TypeMirror rhs) {
```
Comment (wmdietl) [inline src/checkers/inference/DefaultSlotManager.java:601]: For clarity and consistency, let's also use `lhsType`.

```suggestion
            AnnotationLocation location, TypeMirror lhsType, TypeMirror rhsType) {
```

File: src/checkers/inference/DefaultSlotManager.java [hunk 2]
```diff
@@ -612,7 +609,7 @@ public ArithmeticVariableSlot createArithmeticVariableSlot(
         }
 
         // create the arithmetic var slot if it doesn't exist for the given location
-        TypeKind kind = getArithmeticResultKind(lhsAtm, rhsAtm);
+        TypeKind kind = getArithmeticResultKind(lhs, rhs);
```
Comment (wmdietl) [inline src/checkers/inference/DefaultSlotManager.java:612]:
```suggestion
        TypeKind kind = getArithmeticResultKind(lhsType, rhsType);
```

File: src/checkers/inference/SlotManager.java [hunk 3]
```diff
@@ -155,7 +155,7 @@
      * @return the ArithmeticVariableSlot for the given location
      */
     ArithmeticVariableSlot createArithmeticVariableSlot(
-            AnnotationLocation location, AnnotatedTypeMirror lhsAtm, AnnotatedTypeMirror rhsAtm);
+            AnnotationLocation location, TypeMirror lhs, TypeMirror rhs);
```
Comment (wmdietl) [inline src/checkers/inference/SlotManager.java:158]:
```suggestion
            AnnotationLocation location, TypeMirror lhsType, TypeMirror rhsType);
```

Comment (wmdietl) [inline src/checkers/inference/SlotManager.java:158]: The names/comments in the javadoc above also need to be updated.

File: src/checkers/inference/model/ArithmeticConstraint.java [hunk 4]
```diff
@@ -57,6 +57,16 @@ public static ArithmeticOperationKind fromTreeKind(Kind kind) {
                     return OR;
                 case XOR:
                     return XOR;
+                case PLUS_ASSIGNMENT:
```
Comment (wmdietl) [inline src/checkers/inference/model/ArithmeticConstraint.java:60]: Can you add a comment somewhere, why the `_ASSIGNMENT` versions can be handled the same as the non-assignment versions?

File: src/checkers/inference/model/ArithmeticConstraint.java [hunk 5]
```diff
@@ -57,6 +57,16 @@ public static ArithmeticOperationKind fromTreeKind(Kind kind) {
...
+                    return PLUS;
```
Comment (wmdietl) [inline src/checkers/inference/model/ArithmeticConstraint.java:61]: That the two versions are handled the same would be clearer if you used fall-through cases:

````
case PLUS:
case PLUS_ASSIGNMENT:
    return PLUS;
````

File: src/checkers/inference/model/ArithmeticConstraint.java [hunk 6]
```diff
@@ -57,6 +57,16 @@ public static ArithmeticOperationKind fromTreeKind(Kind kind) {
...
+                case MINUS_ASSIGNMENT:
+                    return MINUS;
+                case MULTIPLY_ASSIGNMENT:
+                    return MULTIPLY;
+                case DIVIDE_ASSIGNMENT:
+                    return DIVIDE;
+                case REMAINDER_ASSIGNMENT:
```
Comment (wmdietl) [inline src/checkers/inference/model/ArithmeticConstraint.java:68]: There are a few more tree kinds where this handles the non-`_ASSIGNMENT` version, but not the `_ASSIGNMENT` version, e.g. `LEFT_SHIFT_ASSIGNMENT`. Should these be handled?
See https://github.com/openjdk/jdk/blob/master/src/jdk.compiler/share/classes/com/sun/source/tree/Tree.java for all kinds

File: src/checkers/inference/DefaultSlotManager.java (hunk 1 above)
Comment (wmdietl) [inline src/checkers/inference/DefaultSlotManager.java:601]: Why are there no calls to this method that need to be changed?

File: src/checkers/inference/DefaultSlotManager.java [hunk 7]
```diff
@@ -577,14 +577,11 @@ public ExistentialVariableSlot createExistentialVariableSlot(Slot potentialSlot,
 
     /**
      *  Determine the type kind of an arithmetic operation, based on Binary Numeric Promotion in JLS 5.6.2.
-     * @param lhsAtm atm of left operand
-     * @param rhsAtm atm of right operand
+     * @param lhsType type of left operand
```
Comment (wmdietl) [inline src/checkers/inference/DefaultSlotManager.java:580]: Could you break this PR into two, one for compound assignments and one for these parameter type changes?

**Review by wmdietl — COMMENTED** [review]
Thanks @d367wang !

File: src/checkers/inference/DefaultSlotManager.java (hunk 1 above)
Comment (d367wang) [reply src/checkers/inference/DefaultSlotManager.java]: Because `ArithmeticVariableSlot`s are only created in the value-inference downstream

**Pull request merged**